aesd-char-driver/main.c

```diff
@@ -0,0 +1,268 @@
+#include <linux/slab.h>
+#include <linux/cdev.h>
+#include <linux/fs.h> // file_operations
+
+#include "aesd-circular-buffer.h"
+#include "aesdchar.h"
+int aesd_major = 0; // use dynamic major
+int aesd_minor = 0;
+
+MODULE_AUTHOR("Chinmay Shalawadi"); /** TODO: fill in your name **/
+MODULE_LICENSE("Dual BSD/GPL");
+
+struct aesd_dev aesd_device;
+
+int aesd_open(struct inode *inode, struct file *filp)
+{
+    struct aesd_dev *dev;
+
+    PDEBUG("open");
+
+    dev = container_of(inode->i_cdev, struct aesd_dev, cdev);
+    filp->private_data = dev;
+
+    return 0;
+}
+
+int aesd_release(struct inode *inode, struct file *filp)
+{
+    PDEBUG("release");
+    /**
+     * TODO: handle release
+     */
+    return 0;
+}
+
+ssize_t aesd_read(struct file *filp, char __user *buf, size_t count,
+                  loff_t *f_pos)
+{
+    ssize_t retval = 0;
+    size_t entry_offset;
+    int buff_count = 0;
+    struct aesd_dev *dev_strcut = (struct aesd_dev *)filp->private_data;
+    struct aesd_buffer_entry *entry;
+    PDEBUG("read %zu bytes with offset %lld", count, *f_pos);
+
+    mutex_lock(&(dev_strcut->lock));
+    entry = aesd_circular_buffer_find_entry_offset_for_fpos(&dev_strcut->data_buffer, *f_pos, &entry_offset);
+    mutex_unlock(&(dev_strcut->lock));
+
+    if (entry == NULL)
+    {
+        printk(KERN_ALERT "NULLLLLL \n");
+        *f_pos = 0;
+        goto end;
+    }
+
+    if ((entry->size - entry_offset) < count)
+    {
+        printk("entry offset -> %ld, size-> %ld", entry_offset, entry->size);
+        *f_pos += (entry->size - entry_offset);
+        buff_count = entry->size - entry_offset;
+    }
+    else
+    {
+        *f_pos += count;
+        buff_count = count;
+    }
+
+    if (copy_to_user(buf, entry->buffptr + entry_offset, buff_count))
+    {
+        retval = -EFAULT;
+        goto end;
+    }
+
+    printk("buff_count %d \n", buff_count);
+    retval = buff_count;
+end:
+    return retval;
+}
+
+ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count,
+                   loff_t *f_pos)
+{
+    ssize_t retval = -ENOMEM;
+
+    // Buffer Pointers for logic
+    char *ker_buff, *potentially_free_memory;
+    struct aesd_buffer_entry entry;
+
+    // Device struct from private data
+    struct aesd_dev *buffer = (struct aesd_dev *)filp->private_data;
+
+    // Variables for write logic
+    int newlineflag = 0, i;
+
+    PDEBUG("write %zu bytes with offset %lld", count, *f_pos);
+
+    // Allocating memory for copy buffer
+    ker_buff = kmalloc(count * sizeof(char), GFP_KERNEL);
+
+    if (!ker_buff)
+        goto err;
+
+    if (copy_from_user(ker_buff, buf, count))
+    {
+        retval = -EFAULT;
+        goto err;
+    }
+
+    // Find new line
+    for (i = 0; i < count; i++)
+    {
+        if (ker_buff[i] == '\n')
+        {
+            newlineflag = 1;
+            break;
+        }
+    }
+
+    // if there's no global buffer allocation
+    if (!buffer->global_buffer_size)
```

Comment:
@neohere97 is this safe to access without a mutex lock? What will happen if another process is writing at the same time here?